### DISCUSSION TITLE
Best practices with multiple developers

### DISCUSSION THREAD

**Discussion by khanna7** [khanna-lab/cadre#37] [Q&A] [answered]
What are some best practices to have multiple persons working on the same repository? I understand that typically one creates pull requests, but, the RA at Brown (Yurui Zhang) and I have some questions on how to make that process work well:

- Yurui and I both typically work on the code in separate branches. After I test my code, I pull it into master.

- Currently, Yurui has made some changes in his branch that is 1 commit ahead but 24 commits behind `master`. 

- First, I want to test the change that he made. Do I just pull his branch and run his tests to make sure that the code is working?

- Second, I want to merge in the additional 1 commit that he has made, but without losing the 24 additional commits that I merged into master while he was working on his branch. What is the best way to do that? After testing, would I just go into master and do:

   ``` git merge Yurui-branch``` 

  and this will add in his commit that I am trying to incorporate? (Or might there be some merge conflicts that might need to be resolved)?

- I am also thinking that I would like to prevent the `master` branch from being edited "directly" by anyone. i.e., any changes that one wants to make should be made in a separate branch, then tested, and then merged into master (as opposed to directly modifying and committing changes to the master code.) Does this make sense as a strategy to protect the master branch? If so, do you have any recommendations on how to do that? (Just agree on a set of rules, or set up some formal mechanisms to prevent modifications? I have looked at the "branch protect" documentation, but I am not sure if the features in that go far enough to do this.) 

- Or, are there other schemes that you might recommend to maintain a working and protected version of the codebase?

- Currently, both Yurui and I (as well as Daniel, Nick and Jonathan) are collaborators on the repository. For persons doing most of the development (e.g., Yurui and myself, is it better to be working on a fork of the repository, than the repository itself?

We'll be grateful for any guidance. 
 
Many thanks,
Aditya and Yurui

**Comment by jozik** [comment, marked as answer]
The gitflow pattern is one option: https://www.atlassian.com/git/tutorials/comparing-workflows/gitflow-workflow
This assumes that the main (or master) branch is used for releases, and the main work is done on a `develop` branch, with `feature` branches coming off and merged back into (after testing) to `develop`.

**Reply by khanna7** [reply]
Thanks @jozik!

**Reply by khanna7** [reply]
(Leaving this unmarked for now in case Daniel or Nick have anything to add.)
